### ISSUE TITLE
Alternative to Introduction to Computer Networking on Langunita

### ISSUE THREAD

**Issue by ceik** [ossu/computer-science#520]
I started the course a little while ago and found it to be very tough to get through. There are zero assignments or practical exercises and the only feedback is via in-video questions, which are pretty basic. I ended up writing notes, so the stuff sticks a bit better, but this course essentially became my chore of the day, where I had to force myself to do at least one video a day. I stopped at the end of week 2.

I don't want to skip this topic however so I was looking for alternatives and only found the following two:

- Georgia Tech's Computer Networking on Udacity: https://eu.udacity.com/course/computer-networking--ud436. This is already listed under Extra's and it doesn't seem to be a direct replacement, because the Stanford course is a lot more basic/low level (please CMIIW). Therefore I don't think it is a valid alternative.
- Computer Networks and the Internet by a small German University on edx: https://www.edx.org/course/computer-networks-internet-kironx-fhlcnx. This seems to cover the same content on the same level as the Stanford course.

Has anyone done the second course I listed above or knows of any other alternatives?

**Comment by hanjiexi** [comment]
I am also interested to hear more about the second course mentioned above. If it provides more exercises and feedback while covering the same content as the Stanford course, it might be a viable replacement.

**Comment by ceik** [comment]
Ok, I just started some other courses, but will most likely give the second course mentioned above a try after finishing my current ones. Will leave feedback here in that case.

**Comment by a1h05** [comment]
> https://www.edx.org/course/computer-networks-internet-kironx-fhlcnx

- "_No playable video sources found._" - is shown instead of videos. 
- No programming assignments. 
=(

The main disadvantage of current networking course is lack of programming assignments. We could supplement Lagunita course with url to assigments from Princeton University COS 461 https://www.cs.princeton.edu/courses/archive/spring18/cos461/ (or directly to github https://github.com/PrincetonUniversity/COS461-Public). 
(Have not tried them myself)

**Comment by szaitseff** [comment]
I've just completed the "**Intro to CompNetworking**" by Stanford on Lagunita this month. It was a rather comprehensive intro course with a steep learning curve in my case. I don't know a good alternative as it's free, it has lots of quizzes to test understanding (and not all of them are 'basic'). The mid-term and final exams are challenging. The course is already rather time consuming, and I think adding programming assignment would make it over-tiresome. Just do it through! The Kurose and Ross textbook with free [Wireshark labs](http://www-net.cs.umass.edu/wireshark-labs/) is a good complementary reference for those who wants more practice.

**Comment by fractalbach** [comment]
Tldr
* Yes, add an alternative if there's a good one.
* No, don't remove or replace this course entirely


In addition to the Wireshark labs from the Kurose textbook, there is also a Socket Programming exercise in python (Kurose edition 6, chapter 2.7)

I found that this course was extremely beneficial.  For me, learning how to program clients/servers/etc. in a "practical way" was much easier after taking the course.

If there's another course that works well for a lot of people, it should be added as an alternative.

But don't replace/remove this course of the list entirely.  I found it highly valuable, educational, and worthwhile, and I'm sure there are others who will appreciate it as well.

On a side note...

For those who left the class earlier because they found it tedious, go back and watch some the videos from week 8 (security)

Those are great!  (Especially the demos) 
=)

**Comment by aryzach** [comment]
I agree. I found this class both inefficient at teaching what it does (no homework), and that it covers kinda the wrong material. I think that at points it gets lost in the weeds. I think a class that took out the 'lost in the weeds' portion and substituted practical application demonstrations would be super helpful. Ex. tutorials/assignments with Wireshark and Mininet, and other net/sec tools. Something that combines an intro to networking ideas, with practical sys admin tools

**Comment by kathgironpe** [comment]
@ceik I'm done with CS 6250 at Georgia Tech. It's alright, but even on its own, insufficient. They link the Stanford course for those who want to understand fundamentals.

**Comment by waciumawanjohi** [comment]
It appears that most people are satisfied with the topics covered by the Stanford course and are interested in finding assignments that augment the material. As @kathgironpe points out, the largest online CS masters program recommends this course as a pre-req. I certainly found the lectures appropriate to prepare me for graduate work.

Some courses that we might expect to be high quality alternatives also use simple quizzes for assessments. See for example [Google's intro to networking on Coursera](https://www.coursera.org/learn/computer-networking). [Cisco has courses](https://www.coursera.org/specializations/networking-basics) with labs that require use of network tools. But my impression is that these courses are more professional development courses for network administrators rather than CS courses on network principles. (I welcome discussion on this)

I have seen 3 sets of Computer Network course assignments that we can assume are high quality (one recommend by @szaitseff, another by @a1h05)

- [Princeton 2019](https://github.com/PrincetonUniversity/COS461-Public)
- [Stanford 2010](https://www.scs.stanford.edu/10au-cs144/)
- [Computer Networking: A Top-Down Approach by Kurose Ross](http://www-net.cs.umass.edu/wireshark-labs/)

Some notes on these:
- The Princeton and Stanford assignments come with test suites, while the Kurose and Ross assignments do not.
- The Princeton assignments are actively maintained, with changes in the repository in the last few months.
- The Kurose and Ross book is a well regarded textbook. It is recommended for both the Stanford and Princeton course. We can assume some degree of alignment of the courses with this text, and thereby with its assignments.
- The Princeton course is more similar to advanced networking courses. The supplemental lectures are the same as Georgia Tech's masters degree. One of the Princeton assignments states that it is modeled from a Stanford assignment, which was an Advanced Networking course.

To determine what work students need to undertake, we should look at the expectations for undergraduate computer science students. [Our curriculum guidelines](https://github.com/ossu/computer-science/blob/dev/extras/curriculum_guidelines.md) include the [Curriculum Guidelines for
Undergraduate Degree Programs in Computer Science](https://www.acm.org/binaries/content/assets/education/cs2013_web_final.pdf) from The Joint Task Force on Computing Curricula Association for Computing Machinery (ACM) IEEE Computer Society. Page 130 describes the Computer Networking learning expectations for undergrads.

One thing to note is how many of the networking topics students are meant to understand at a level of "Familiarity".

_Familiarity: The student understands what a concept is or what it means. This level of
mastery concerns a basic awareness of a concept as opposed to expecting real facility
with its application. It provides an answer to the question “What do you know about
this?”_

By this standard, for most networking topics we can assume that quizzes in the place of labs are acceptable. There are 4 learning outcomes that go beyond familiarity. Below I match these goals with assignments available online.

- Implement a simple client-server socket-based application.
  - [Princeton assignment on creating a server and client](https://github.com/PrincetonUniversity/COS461-Public/tree/master/assignments/assignment1)
- Design and implement a simple reliable protocol.
  - [Stanford assignment creating reliable transport](https://www.scs.stanford.edu/10au-cs144/lab/reliable/reliable.html)
- Compare and contrast fixed and dynamic allocation techniques.
  - [Read the following chapter and respond to the short answer questions.](https://nptel.ac.in/courses/106105080/pdf/M2L7.pdf)
- Compare and contrast current approaches to congestion.
  - [Wireshark Lab on TCP](http://www-net.cs.umass.edu/wireshark-labs/Wireshark_TCP_v7.0.pdf)
  - Alternatives include:
    - https://computer-networks.github.io/sp19/handout/project2.pdf
    - https://www2.cs.duke.edu/courses/spring14/compsci356/labs/lab3/lab3.html
    - http://user.it.uu.se/~perg/course/datakom2/it98/tcpsims_lab.html
    - https://booksite.elsevier.com/9780123850591/Lab_Manual/Lab_10.pdf

Based on the above, I propose that we retain the Stanford Lagunita course as the lectures for Computer Networking. In addition, I propose we specify the 4 above assignments for students.

I will leave this issue open for 1 month for discussion of this proposal. Suggestions of alternative lecture series, alternative assignments, even alternative curricula guidelines are welcome and encouraged. Helpful recommendations will include in depth analysis of the strengths and weaknesses of the materials and/or of the fit of the material to the learning goals.

**Comment by kathgironpe** [comment]
I know a lot of you checking this repo are not graduate school students like myself, and that's even better because you're likely not doing this to get an A. Might as well give you some hints on what it feels like to go through it. It's a top 6 global CS University, and even with that fact, education is really just something you can only give to yourself. I am suggesting that I could have learned a lot more by creating my own curriculum. But I don't regret going through this.

For those who feel insufficient in terms of skills because they're not doing any projects, I don't think you are missing a lot for *that reason.* You are missing more because there aren't a lot of papers linked to courses like Stanford's Introduction to Computer Networking. 

Over 3 months, I have read  over **37 Computer Network books,** and more than 40 research papers on Computer Networks. It feels like a lot, but actually, it's NOT. What's worse: when it comes to the actual exam, it's my lack of CS fundamentals and formal education on Distributed Computing that made me scratch my head. I can't believe how seemingly simple some questions were to those with background, but I didn't know what they meant. I leaped from a top 200+ University to a top 6 in CS, and knew something like this could happen. I just get shocked about the terminology that I never encountered on the transcripts. 

It must be fun to just create your own curriculum. 
What is a sufficient introduction to Computer Networking?

I would still list the Stanford Introduction to Computer Networking as a number one resource.
And probably add:

[CS 6250 at GT](https://classroom.udacity.com/courses/ud436)
[CMU's Computer Networks' Syllabus](https://www.cs.cmu.edu/~15744/syllabus.html)

For real world work, you definitely need a lot more than this and CompTIA courses. AWS certificate courses help.

If you're doing this for interview preparation, I can assure you that Computer Networking questions are common. You might not expect companies like AirBnB to ask you to convert an [I.P range to CIDR but they do](https://github.com/allaboutjst/airbnb/blob/master/src/main/java/ip_range_to_cidr/IPRangetoCIDR.java). If you hated these kind of questions in school, you might get them when you apply for work.

**Comment by omaraloraini** [comment]
[14-740](http://www.ini740.rocks/F18/index.html#schedule)

Graduate level with lots of reading, it seems not to different from the Stanford course in terms of content.

**Comment by waciumawanjohi** [comment]
There is a pull request which addresses this issue. Accepting comments on the PR for the next week.

https://github.com/ossu/computer-science/pull/590
